[MIG] account_move_name_sequence: Migration to 18.0

CLaurelB opened the pull request:
Supersed: https://github.com/OCA/account-financial-tools/pull/1954

alexis-via committed:
Add module account_move_name_sequence

This module restores the good old behavior where journal entry numbers
were generated from a sequence configured on the journal.

alexis-via committed:
sequence is now required on journals

Add post-install script to create a sequence for all existing journals
Update README accordingly

moylop260 committed:
[REF] account_move_name_sequence: Add number_next_actual and date range lines data based on current moves

moylop260 committed:
[FIX] ir_sequence: Fix creating date_range_seq with range_month and range_day

More info about:
 - https://github.com/odoo/odoo/pull/91019

TODO: Revert this commit after it is merged

moylop260 committed:
[REF] account_move_name_sequence: Discard journals already configured

moylop260 committed:
[FIX] account_move_name_sequence: Fix sequence domain to assign only if it is empty

moylop260 committed:
[REF] account_move_name_sequence: Moves with name '/' can not be posted

After remove required=True for journal.sequence_id field it is possible to post an invoice with misconfigured journal with empty sequence

So, this constraint will raise an error for this kind of cases since that using '/' could raise the unique constraint for all other moves

moylop260 committed:
[FIX] account_move_name_sequence: Fix required journal fields in view

The required flag was wrong for sequence_id and refund_sequence_id

So, it was not possible to store any change in journal for journal different to sale and purchase

moylop260 committed:
[REF] account_move_name_sequence: Add maintainers to myself

moylop260 committed:
[REF] account_move_name_sequence: Adding concurrency cases in README

frahikLV committed:
[FIX] account_move_name_sequence: Exclude no_gap sequences from _is_end_of_seq_chain

frahikLV committed:
[REF] account_move_name_sequence: Use Odoo native methods + Typos

frahikLV committed:
[MIG] account_move_name_sequence: Migration to 15.0

petrus-v committed:
[FIX] account_move_name_sequence: Use account move date to compute prefix

In case you want name your invoice YYYY-MM-SEQ (ie: 2022-07-00001)
where:
 * YYYY: is the account move year
 * MM: is the account move month
 * SEQ: is a numerical sequence that is continue along the fiscal year
   assuming fiscal year is over two years (ie: from july to june next year)

Before this commit the sequence prefix use now() to be compute but the
range is selected with the account move date.

This commit make consistency computing prefix with the account
move date as well.

So account move manage the first janunary for the last day of
the previous year will properly use the account move date.

Co-authored-by: Alexis de Lattre <alexis.delattre@akretion.com>

payen000 committed:
[REF] account_move_name_sequence: Added concurrency tests related to sequence

A demo ir_sequence record with 'standard' implementation was
assigned to the payment journal so that no concurrency errors arise
from using 'no gap' sequences when attempting to create multiple payments.

RodrigoBM committed:
[MIG] account_move_name_sequence: Migration to 16.0

RodrigoBM committed:
[FIX] account_move_name_sequence: return sequence names do not add a space between name and returns.

RodrigoBM committed:
[FIX] account_move_name_sequence: issue #1465 vals get refund_sequence true if value not get in vals

fernandahf committed:
[FIX] account_move_name_sequence: call flush before to post the moves

RodrigoBM committed:
[FIX] account_move_name_sequence: call flush before _fetch_duplicate_supplier_reference method
same issue https://github.com/OCA/account-financial-tools/issues/1501
this fix not working for v16 https://github.com/OCA/account-financial-tools/pull/1514

moylop260 committed:
[REF] account_move_name_sequence: Remove concurrency outdated cases from the readme

RodrigoBM committed:
[FIX] account_move_name_sequence: With the "account_move_name_sequence" module the "_get_last_sequence" method does not have to propagate the with_prefix parameter.  The sequence_prefix parameter will not be completed and will give error as it is False in this line of code. https://github.com/OCA/OCB/blob/16.0/addons/account/models/sequence_mixin.py#L169

RodrigoBM committed:
[FIX] account_move_name_sequence: Now you can unlink moves without check sequence fix: https://github.com/odoo/odoo/commit/6112f4f59aeb5dc5a3b698edd861b19f25da7017

badbole committed:
Added translation using Weblate (Croatian)

remi-filament committed:
[FIX] account_move_name_sequence: computed fields raising warning

petrus-v committed:
[FIX] account_move_name_sequence: avoid fallback on odoo sequence name

In case an alias is configured to fecth email to generate
account entries. the new_message method directly call the
_compute_name on account move which call the legacy code
provide by account odoo module which should be avoid here.

To reproduce:

* configure an alias on your purchase journal
* send an email to that alias
* the invoice is draft with a number

to mitigate this behavior the suggested patch overwritte
the former method '_compute_name' and call the new implemented
method '_compute_name_by_sequence'.

sysadminmatmoz committed:
Added translation using Weblate (Slovenian)

remi-filament committed:
Added translation using Weblate (French)

Gelo-fl committed:
Added translation using Weblate (Spanish)

Commit:
Translated using Weblate (French)

Currently translated at 100.0% (20 of 20 strings)

Translation: account-financial-tools-16.0/account-financial-tools-16.0-account_move_name_sequence
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-16-0/account-financial-tools-16-0-account_move_name_sequence/fr/

astirpe committed:
[FIX] account_move_name_sequence: do not create duplicated sequences

hussain committed:
Added translation using Weblate (Arabic)

andreagidaltig committed:
[REF] account_move_name_sequence: rename hooks file

Rename hooks file and use post_init_hook method to create journal sequences
after module installation.

BT-anieto committed:
[FIX] account_move_name_sequence: call _inverse_name after computing the name

The core compute function for the name is calling at the end the function self._inverse_name(),
which updates the payment_reference when required.
This code was currently missing, causing the payment_reference not being properly computed,
for example, when using QR-Bills.

mide-dynapps committed:
[FIX] account_move_name_sequence: redefine _onchange_journal_id

In v16, _onchange_journal_id is defined to update name when journal_id is modified
During the creation of new account.move, this onchange is triggered and called the standard _compute_name method

How to reproduce issue:
* Create new invoice
=> Name is defined instead of to be 'Draft'

andreagidaltig committed:
[IMP] account_move_name_sequence: apply guidelines to name view files

Update name of view files to follow guidelines.

andreagidaltig committed:
[IMP] account_move_name_sequence: split too long context definition

Split the context definition in sequence_id and refund_sequence_id
fields into multiple lines to improve readability.

moitabenfdz committed:
[IMP] account_move_name_sequence: pre-commit stuff

moitabenfdz committed:
[MIG] account_move_name_sequence: Migration to 17.0

Commit:
[UPD] Update account_move_name_sequence.pot

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (20 of 20 strings)

Translation: account-financial-tools-17.0/account-financial-tools-17.0-account_move_name_sequence
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-17-0/account-financial-tools-17-0-account_move_name_sequence/it/

moylop260 committed:
[IMP] account_move_name_sequence: pre-commit auto fixes

moylop260 committed:
[FW] account_move_name_sequence: Forward-Port of 14.0 to 15.0

ljsalvatierra-factorlibre committed:
[REF] account_move_name_sequence: Check invoice name based on journal entry sequence

moylop260 committed:
[FW] account_move_name_sequence: Forward-Port of 15.0 to 16.0

moylop260 committed:
[IMP] account_move_name_sequence: pre-commit auto fixes

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

Tardo committed:
fix: account_move_name_sequence: tests: Listen to exceptions for each thread individually"

OCA-git-bot committed:
[BOT] post-merge updates

moylop260 committed:
[REF] account_move_name_sequence: Add unittest to create concurrency invoices

Add demo data with standard implementation sequences

In order to test locks issues without changing data (then reverting)

Commit:
[UPD] Update account_move_name_sequence.pot

OCA-git-bot committed:
[BOT] post-merge updates

OCA-git-bot committed:
[BOT] post-merge updates

weblate committed:
Update translation files

Updated by "Update PO files to match POT (msgmerge)" hook in Weblate.

Translation: account-financial-tools-17.0/account-financial-tools-17.0-account_move_name_sequence
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-17-0/account-financial-tools-17-0-account_move_name_sequence/

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (22 of 22 strings)

Translation: account-financial-tools-17.0/account-financial-tools-17.0-account_move_name_sequence
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-17-0/account-financial-tools-17-0-account_move_name_sequence/it/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (22 of 22 strings)

Translation: account-financial-tools-17.0/account-financial-tools-17.0-account_move_name_sequence
Translate-URL: https://translation.odoo-community.org/projects/account-financial-tools-17-0/account-financial-tools-17-0-account_move_name_sequence/es/

CLaurelB commented:
@luisg123v could you review, please?

thienvh332 commented on account_move_name_sequence/models/account_journal.py:
```diff
@@ -0,0 +1,251 @@
+                and journal.refund_sequence_id == journal.sequence_id
+            ):
+                raise ValidationError(
+                    _(
```
```suggestion
                    self.env._(
```

thienvh332 replied:
Similar to some other positions

CLaurelB replied:
Suggestion applied.

moylop260 replied:
FYI New pylint adds this new check

Maybe we should update the pylint version from template in order to update the projects and avoid creating a manual review for this

cc @sbidoul

sbidoul replied:
I'm open to a PR to upgrade the pylint-odoo version on oca/oca-addons-repo-template.

This should be a PR that asks a copier question to let the PSC choose the version. This is because we cannot simply upgrade the version on all repos and all branches at once, as a mass update would make many branches red.

thienvh332 reviewed (approved):
Just nitpicking :wink:

thienvh332 commented on account_move_name_sequence/models/account_journal.py:
```diff
@@ -129,7 +131,7 @@ def _prepare_sequence_current_moves(self, refund=False):
         )
         if not last_move:
-            _logger.warning("%s %s", msg_err, "No moves found")
+            _logger.debug("%s %s", msg_err, "No moves found")
```
I wonder why this change? I think it would be more reasonable for us to WARNING?

CLaurelB replied:
Change reverted.

thienvh332 replied:
Hehe, You can use `assertLogs` to intercept and check for warnings. It will help you pass the tests below :wink:

CLaurelB replied:
Since the issue was in the `post_init_hook`, I added a context validation to mute the warning when desired.

moylop260 replied:
No mute logger using context only for test

Better using mute logger from the test

CLaurelB replied:
@moylop260 replaced by mute_logger.

luisg123v commented on account_move_name_sequence/models/account_journal.py:
```diff
@@ -30,7 +30,8 @@ class AccountJournal(models.Model):
         domain="[('company_id', '=', company_id)]",
-        help="This sequence will be used to generate the journal entry number for refunds.",
+        help="This sequence will be used to generate "
+        "the journal entry number for refunds.",
```
Why this change?

CLaurelB replied:
Because of the pre-commit suggested change.

luisg123v replied:
Is that done only after migrating?

CLaurelB replied:
After running pre-commit, this change is flagged as an error with a suggested fix, rather than being applied automatically. I included the suggested fix in the migration commit.

luisg123v commented on account_move_name_sequence/tests/test_account_move_name_seq.py:
```diff
@@ -5,65 +5,70 @@
+        cls.date = datetime.now()
+
+        cls.journals = cls.misc_journal | cls.purchase_journal | cls.sales_journal
+        with patch("odoo.models.BaseModel._validate_fields"):
```
Since it's not a trivial change, we could add a coment above explaining what's this for. This because patches are not functionally possible, so I think we should provide an explanation in such cases.

Same in similar cases were patches are applied.

CLaurelB replied:
Done.

luisg123v commented on account_move_name_sequence/tests/test_sequence_concurrency.py:
```diff
@@ -85,31 +94,38 @@ def _create_payment_form(self, env, ir_sequence_standard=False):
-            # TODO: Delete payment and journal
+            moves = env["account.move"].with_context(force_delete=True).browse(move_ids)
+            payments = moves.payment_ids
+            moves_without_payments = moves - payments.mapped("move_id")
```
`mapped` not required

CLaurelB replied:
Removed.

luisg123v commented on account_move_name_sequence/tests/test_sequence_concurrency.py:
```diff
@@ -85,31 +94,38 @@ def _create_payment_form(self, env, ir_sequence_standard=False):
+                payments.unlink()
+            if moves_without_payments:
+                moves_without_payments.button_draft()
+                moves_without_payments.unlink()
```
Why not removing all moves instead of only the ones without payments?

CLaurelB replied:
Removing the payments already removes the entries related to them. Therefore, it is only necessary to remove the other moves that were not created from a payment.

luisg123v commented on account_move_name_sequence/tests/test_sequence_concurrency.py:
```diff
@@ -85,31 +94,38 @@ def _create_payment_form(self, env, ir_sequence_standard=False):
         self, deadlock_timeout, payment_first=False, ir_sequence_standard=False
     ):
-        odoo.registry(self.env.cr.dbname)
+        Registry(self.env.cr.dbname)
```
I think this does nothing.

CLaurelB replied:
Removed.

luisg123v commented on account_move_name_sequence/tests/test_sequence_concurrency.py:
```diff
@@ -344,6 +373,7 @@ def test_sequence_concurrency_95_pay2inv_inv2pay(self):
                 args=(deadlock_timeout, False, True),
                 name="Thread invoice payment",
             )
+            self.env.registry.enter_test_mode(self.env.registry.cursor())
```
I think this should be done in the `setUpClass` method, along with the `addClassCleanup` to exit it.

CLaurelB replied:
Refactored to have also the new cursors in the `SetUpClass`.

luisg123v commented:
/ocabot migration account_move_name_sequence

luisg123v reviewed (approved):
LGTM :+1: 

@moylop260 you have more context about this module. What do you think?

CC @desdelinux

moylop260 commented:
Hi @CLaurelB 

I'm a little confused about what change is required for migration and what change is a improvement.

Could you split the commits to:
1) autofixes e.g. running pre-commit and autofixed
2) manual lints fixes e.g. pre-commit red and without autofixed
3) Migration and only migration changes (e.g. self.env._)
4) Improvements (more coverage, any change not related with migration or lints), refactorings (test refactoring)

I think this last commit could be backported to other versions but currently is merged with the migration script, but increase coverage or refactoring is not part of migration and it is not easy to migrate and review

moylop260 commented on account_move_name_sequence/demo/account_journal_demo.xml:
```diff
@@ -4,15 +4,15 @@
         <field name="code">SSJD</field>
         <field name="type">sale</field>
-        <field name="refund_sequence">True</field>
+        <field name="refund_sequence" eval="True" />
```
I was not aware this change for v18.0

Odoo v18.0 doesn't support the old way without eval?


Do you have the reference to this change in odoo?

CLaurelB replied:
This is not a migration change; it would be in the improvements commit since, with `eval`, the expression is evaluated as a Python expression, while without it, it is treated as a string. It also helps differentiate string fields from non-string fields.

StefanRijnhart replied:
For the record, this is no longer necessary after this change from 2016: https://github.com/odoo/odoo/commit/dc18056fd01848c51a8dfe83f22bcc268aaefea4

CLaurelB committed:
[IMP] account_move_name_sequence: pre-commit auto fixes

CLaurelB committed:
[IMP] account_move_name_sequence: manual lints fixes

Co-authored-by: Bert Van Groenendael <bert.vangroenendael@dynapps.be>

CLaurelB committed:
[MIG] account_move_name_sequence: Migrate to v18.0

Changelog:
- Replace `_()` by `self.env._()`.
- Replace `_fetch_duplicate_supplier_reference` by
  `_fetch_duplicate_reference` to include both sale and purchase
  documents.
- Use `mute_logger` to suppress warnings when executing the
  `_prepare_sequence_current_moves` method without moves in the journals
  from the `post_init_hook`.
- Patch the `_validate_fields` and `_hash_moves` methods in tests to
  simulate the same behavior as if the journal entries were unlocked.
  This is necessary because, in the pipeline, the
  `account_journal_restrict_mode` module is installed, which locks
  posted entries, preventing them from being set to draft or deleted
  afterward.
- Enter test mode when executing threads in the
  `test_sequence_concurrency_95_pay2inv_inv2pay` test to avoid locks.
- Replace `_clean_moves` by `_clean_moves_and_payments`, since entries
  created from a payment cannot be deleted on their own and must be
  deleted along with the payment.

Co-authored-by: Bert Van Groenendael <bert.vangroenendael@dynapps.be>
Co-authored-by: oury.balde <oury.balde@xcg-consulting.fr>

CLaurelB committed:
[IMP] account_move_name_sequence: refactor tests and improve field handling

Changelog:
- Use Command syntax for better record manipulation.
- Use `eval` for non-string field values to ensure proper evaluation.
- Refactor tests to use `SetUpClass`, adding cleanup methods, new
  cursors, and test mode setup within it.
- Add additional tests for sequence prefix and chain to increase
  coverage.

CLaurelB commented:
> Hi @CLaurelB
> 
> I'm a little confused about what change is required for migration and what change is a improvement.
> 
> Could you split the commits to:
> 
> 1. autofixes e.g. running pre-commit and autofixed
> 2. manual lints fixes e.g. pre-commit red and without autofixed
> 3. Migration and only migration changes (e.g. self.env._)
> 4. Improvements (more coverage, any change not related with migration or lints), refactorings (test refactoring)
> 
> I think this last commit could be backported to other versions but currently is merged with the migration script, but increase coverage or refactoring is not part of migration and it is not easy to migrate and review

Hello @moylop260 

The commit has been split into the suggested ones.

Could you review again, please?

Regards.

moylop260 commented:
/ocabot merge nobump
